vllm/core/scheduler_v2.py

```diff
@@ -0,0 +1,370 @@
+            num_tokens = request.num_tokens - num_computed_tokens
+            num_tokens = min(num_tokens, token_budget)
+            assert num_tokens > 0
+            new_block_ids = self.kv_cache_manager.allocate_slots(
+                request, num_tokens, computed_block_ids)
+            if new_block_ids is None:
+                # The request cannot be scheduled.
+                break
+            request.num_computed_tokens = num_computed_tokens
+
+            self.waiting.popleft()
+            self.running.append(request)
+            if request.status == RequestStatus.WAITING:
+                scheduled_new_reqs.append(request)
+            elif request.status == RequestStatus.PREEMPTED:
+                scheduled_resumed_reqs.append(request)
+            else:
+                assert False, f"Invalid request status: {request.status}"
+
+            req_to_new_block_ids[request.request_id] = (computed_block_ids +
+                                                        new_block_ids)
+            num_scheduled_tokens[request.request_id] = num_tokens
+            token_budget -= num_tokens
+            request.status = RequestStatus.RUNNING
+
+        # Check if the scheduling constraints are satisfied.
+        total_num_scheduled_tokens = sum(num_scheduled_tokens.values())
+        assert total_num_scheduled_tokens <= self.max_num_scheduled_tokens
+        assert token_budget >= 0
+        assert len(self.running) <= self.max_num_running_reqs
+        assert (len(scheduled_new_reqs) + len(scheduled_resumed_reqs) +
+                len(scheduled_running_reqs) == len(self.running))
+
+        # Construct the scheduler output.
+        new_reqs_data = [
+            NewRequestData.from_request(req,
+                                        req_to_new_block_ids[req.request_id],
+                                        req.num_computed_tokens)
+            for req in scheduled_new_reqs
+        ]
+        resumed_reqs_data = [
+            ResumedRequestData.from_request(
+                req, req_to_new_block_ids[req.request_id],
+                req.num_computed_tokens) for req in scheduled_resumed_reqs
+        ]
+        running_reqs_data = [
+            RunningRequestData.from_request(
+                req, req_to_new_block_ids[req.request_id],
+                req.num_computed_tokens) for req in scheduled_running_reqs
+        ]
+        preempted_req_ids = {req.request_id for req in preempted_reqs}
+        scheduler_output = SchedulerOutput(
+            scheduled_new_reqs=new_reqs_data,
+            scheduled_resumed_reqs=resumed_reqs_data,
+            scheduled_running_reqs=running_reqs_data,
+            num_scheduled_tokens=num_scheduled_tokens,
+            total_num_scheduled_tokens=total_num_scheduled_tokens,
+            preempted_req_ids=preempted_req_ids,
+            finished_req_ids=self.finished_req_ids,
+            aborted_req_ids=self.aborted_req_ids,
```

Comment:
Nit: Maybe add a comment to distinguish these two fields with other fields.
```suggestion
            # These two fields are existing states in the scheduler instead of newly scheduled in this step.
            finished_req_ids=self.finished_req_ids,
            aborted_req_ids=self.aborted_req_ids,
```